README
Welcome to the miniDB library.

This library was created to have a small fast isam/btree library
for record keybased access. There is no sql interface available or planned.

Features:
- small footprint (less than 26 Kb)
- multiple indexes on a datafile (max 10)
- multiple fields in a index (max 8)
- autoincrement keys
- datafile is ascii based
- cmdline interface for usage in scripts
- client/server or standalone usage
- database logging
- single executable

For more information about btrees :
Book : Sorting and Searching Algorithms by Thomas Niemann, Portland, Oregon  epaperpress.com
* Website: http://www.oopweb.com/Algorithms/Documents/Sman/VolumeFrames.html


Included files:
- bt.c			all btree routines
- mdb_lib.c		all mdb isam routines	
- mdb_client.c		mdb wrapper for client/server use
- mdb_net.c		all mdb network routines	
- mdb.c			miniDB standalone with a cmdline interface
- mdbserver.c	 	miniDB server daemon	
- mdbfilter.c           filter program for mdb output
- mdb2xml.c             output from mdb is converted to xml

The isam datafile is an ascii based file, so that it still can be parsed
using standard utilities (like grep, awk, ..). 
All valid records in the datafile start with the letter R, so to
extract all the records : grep "^R" demo.dat

-------------------------------------------------------------------------
There is a simple cmdline interface called 'mdb' which you can use 
standalone or as client/server program to access the data.

mdb syntax:

Syntax: mdb [-b<idstring>] [-r<norows>] [-k<keyno>] [-h <host>] [-p <port>]
                           <tablename> <cmd> [<key> [<data>]]
 Options:  -b  assign a stringid for recognising empty fields
               used when updating a record
           -f  show field names before data
           -k  select key
           -r  retrieve .. number of rows
           -h  hostname (only in case of client/server)
           -p  portno (default 7221)
           -u  used fields

       cmd = f,l   (first,last)
             p,e,n (previous, equal, next)
             a,u,d (add, update, delete)
	     s     (search)
	     r     (retrieve record 1..n)
	     B,C,D (reBuild, create or drop index)
	     L     (database logging on/off)

mDB always want to have a : filename & cmd

Examples: 

New file creation:
   Ex. mdb customer c 'custid:n6|seak:c8|name:a30|place:a30|_idx:1|_idx:2:1|' 

   Will create a new isamfile called customer with 2 indexes.
   each field is seperated by '|' and each field specifies a type and length
   so: custid:n6  =  numeric, length 6 
       seak:c8    =  ascii,   length 8
       name:a30   =  ascii,   length 30
   types:  a - ascii field
	   c - ascii field, in keys it is uppercased and stripped from spaces
               ex. 'John Doe' will be used in an index as 'JOHNDOE' but
                   in the data it will still be 'John Doe' 
           n - numeric field
	   i - numeric field, when 0 it is given an autoincrement value
	   f - float field (syntax f8.2)

   All data is saved as ascii in the isam file with '|' as a seperator.


Adding data:
   Ex. mdb customer a '1201|doe|John Doe|New York|' 
       1|1201|doe|John Doe|New York|

   Adds a new record to file customer and its indexes.
   fieldvalues seperated by '|', autoincrement fields can be 0
   Will output the added record to standard out with its recordnumber prefixed
   and with the generated autoincement values.

   mdb -u'place|name|custid|seak' customer a '1201' 'London|John Doe|1201|doe'

   The -u flag maps the record fields from the last parameter (record values)

Retrieving:
   Ex. mdb          customer e '1201'  retrieve exact matching key '1201'
       mdb          customer n '1201'  retrieve the next key
       mdb          customer p '1201'  retrieve the previous key
       mdb          customer f         retrieve the first record using index 1
       mdb -k2      customer l 	       retrieve the last  record using index 2
       mdb -k2 -r10 customer s 'do'    search on index 2 as >= 'do' and 
				       retrieve 10 rows 
       mdb -r-10    customer s 'do'    same, but does the previous records

Updating:
   ex. mdb -b'#' customer u '1201' '#|#|London|' 

   Will retrieve indexkey '1201' (index 1) and update its field contents, i
   leaving the '#' fields to the original values.
   Note: Will update only 1 record.

      mdb -b'#' -u'place|name' customer u '1201' 'London|Doe'

   The -u flag will specify the last parameter and internally it will be build
   like: '#|#|Doe|London|' 

Deleting:
   Ex. mdb customer d '1201' 

   Will delete the record with indexkey '1201' (index 1) .


Building, adding and deleting indexes:
   Ex. mdb -k2 customer D  		Delete index 2
       mdb -k2 customer C '2:1' 	Add index 2 with field 2 and 1 (seak/id)
       mdb -k2 customer B 		Rebuilds index 2
       mdb -k0 customer B 		Rebuilds all indexes

Enable/disable logging:
   Ex. mdb customer L on		Enable file logging
   Ex. mdb customer L off		Disable file logging


   With filelogging all inserts,updates and deletes are logged into a 
   seperate logfile. With this logfile changes can be traced and for
   instance an incremental backup can be made or changes can be reversed.
   When updates are logged the old and new record values are being logged.

Information:
   Ex. mdb customer i			gives some mdb information 


Extra programs:

mdbfilter - will filter the output of mdb
            ex. mdb -f -r10 demo s data|mdbfilter -f "name|id|"

mdb2xml   - output from mdb will be outputted as xml
            ex. mdb -f -r10 demo s data|mdbxml -p "tbl_"

mdb2json  - output from mdb will be outputted as JSON
            ex. mdb -f -r2 demo s '12344'|mdb2json -p "tbl_" -r

JSON output example:

{ "tbl_Result": [
     { 
	 "tbl_recno": 11344,
	 "tbl_id": 12344,
	 "tbl_seak": "infirm",
	 "tbl_name": "infirm",
	 "tbl_street": "street 12344",
	 "tbl_postal": "ZIP 1234",
	 "tbl_place": "nfirm"
     },{ 
	 "tbl_recno": 11345,
	 "tbl_id": 12345,
	 "tbl_seak": "affirmed",
	 "tbl_name": "affirmed",
	 "tbl_street": "street 12345",
	 "tbl_postal": "ZIP 1234",
	 "tbl_place": "ffirmed"
     }
     ],
  "tbl_NoRows": 2 
}



CHANGES
Revision 2.4  2010/07/09 18:19:11
Added mdb2json to produce JSON style output  
(Thanks to Mark Constable for suggestion and help with MIT license)

Revision 2.3  2010/07/07 18:01:30
Change the license terms to MIT which is a more standard license
Adapted some code so it compiles without warnings on gcc 4.4
Formatted code to K&R style

Revision 2.2  2010/02/26 17:10:30
Change the license terms in the programcode 
to match the one on the website.
No code changes were done.

Revision 2.1  2003/10/09 21:04:30
Replaced BPLUS package with new btree library
Added extra programs mdbfilter and mdb2xml
More options and commands
mdb program is now 26K large (small)

Revision 1.0  2003/03/23 20:16:37
Initial revision




src/mdb.c
/* -------------------------------------------------------------------------- *
 * Description  : miniDB interface
 *
 * Revision 2.1 : added cmd for rebuildindex function
 * Revision 2.0 : Adapted for new mdb_lib functions
 * Revision 1.0 : Initial revision
 *
 * ------------------------------------------------------------------------- *
 *
 * Copyright (c) 2010 Hans Harder
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *
 * ------------------------------------------------------------------------- */
#include <stdio.h>
#include <getopt.h>
#include "mdb_lib.h"
#include "mdb_func.h"

int 	NOROWS=1;		/* no rows to retrieve */


void syntax() {
    puts("Syntax: mdb [-b<idstring>] [-r<norows>] [-k<keyno>] [-h <host>] [-p <port>]");
    puts("                           <tablename> <cmd> [<key> [<data>]]");
    puts(" Options:  -b  assign a stringid for recognising empty fields");
    puts("               used when updating a record");
    puts("           -f  show field names before data");
    puts("           -k  select key");
    puts("           -r  retrieve .. number of rows");
    puts("           -h  hostname (only in case of client/server)");
    puts("           -p  portno (default 7221)");
    puts("           -u  used fields");
    puts("");
    puts("       cmd = f,l   (first,last)");
    puts("             p,e,n (previous, equal, next)");
    puts("             a,u,d (add, update, delete)");
    puts("	     s     (search)");
    puts("	     r     (retrieve record 1..n)");
    puts("	     B,C,D (reBuild, create or drop index)");
    puts("	     L     (database logging on/off)");
    puts(" Examples: ");
    puts("   mdb customer c 'custid:n6|seak:c8|Name:a30|place:a30|_idx:1|_idx:2:1|' ");
    puts("   mdb customer a '1201|doe|John Doe|New York|' ");
    puts("   mdb customer e '1201' ");
    puts("   mdb -k2 -r10 customer s 'do' ");
    puts("   mdb -b'#' customer u '1201' '#|#|London|' ");
    puts("   mdb -k2 customer D  ");
    puts("   mdb -k2 customer C '2:1' ");
    exit(1);
}

void do_error(char *s) {
    puts(s);
    exit(1);
}

/*
 * rearrange supplied data in the correct recordlayout
 * ex. supplied data and layout : "Doe|John|1234"  "lastnm|firstnm|id"
 *     recordlayout       : "id|firstnm|lastnm|place|country"
 *     rearranged output  : "1234|John|Doe|#|#"
 */
void rebuilddata(src,uflds,flds,deflt,dst)
char *src,*uflds,*flds,*deflt,*dst;
{
    char fldval[257],tmp[61];
    int  i,j;

    dst[0]=0;
    /* for each recordfield, find a supplied value
     * skip first field in recorddef, because its the recno */
    for (j=2; getfield(j,flds,fldval,60,'|')!=0; j++) {
        if (fldval[0]==0) break;
        /* check if supplied by user */
        for (i=1; getfield(i,uflds,tmp,60,'|')!=0; i++)
            if (tmp[0]==0 || strcmp(fldval,tmp)==0)  break;
        if (tmp[0]) {	/* if found, fill with the corresponding data */
            getfield(i,src,fldval,256,'|');
            strcat(dst,fldval);
        } else		/* else use the default value */
            strcat(dst,deflt);
        strcat(dst,"|");
    }
}

int addrecord(fn,data,dataflds,fldnms)
char *fn,*data,*dataflds,*fldnms;
{
    if (dataflds[0]) {
        char dta[513];
        rebuilddata(data,dataflds,fldnms,"",dta);
        return mDB_add(fn,dta,0);
    } else
        return mDB_add(fn,data,0);
}

int  main(argc, argv)
int  argc;
char *argv[];
{
    char fn[81];
    char keystr[513],fldnms[256];
    char EMPTYID[11]="#";
    char UPDFLDS[256];
    int  FIELDS, st,ch, keylen, idxno;

    FIELDS=0;
    idxno=1;
    st=0;
    UPDFLDS[0]=0;
    while ((ch=getopt(argc,argv,"b:fh:p:k:r:u:"))!=-1) {
        switch (ch) {
        case 'b' :
            strcpy(EMPTYID,optarg);
            break;
        case 'f' :
            FIELDS=1;
            break;
        case 'k' :
            idxno=atoi(optarg);
            break;
        case 'r' :
            NOROWS=atoi(optarg);
            break;
        case 'h' :
            strcpy(mDBhostname,optarg);
            break;
        case 'p' :
            mDBportno=atoi(optarg);
            break;
        case 'u' :
            strcpy(UPDFLDS,optarg);
            break;
        default  :
            syntax();
            break;
        }
    }
    argc -=optind;
    argv +=optind;
    if (argc<2)  syntax();

    if (mDB_init()==0) {
        puts("Database initialisation unsuccessfull !");
        exit(-1);
    }

    sprintf(fn,"data/%s",argv[0]);
    strcpy(keystr,(argc>2)?argv[2]:"");
    keylen=strlen(keystr);
    if (strchr("ck",argv[1][0])) {
        switch (argv[1][0]) {
        case 'c' :
            if (mDB_create(fn,keystr)==0) do_error("create error");
            break;
        case 'k' :
            if (mDB_kill(argv[0])==0) do_error("access denied");
            break;
        }
    } else {
        if (mDB_open(fn)==0) {
            printf("Unknown databasefile [%s]\n",fn);
            exit(-2);
        }
        mDB_fields(fn,fldnms);
        if (FIELDS) puts(fldnms);
        switch (argv[1][0]) {
        case 'a' :
            if (strcmp(keystr,"-")==0) {
                while (fgets(keystr,512,stdin)!=NULL) {
                    strtrim(keystr);
                    if (keystr[0])
                        st=addrecord(fn,keystr,UPDFLDS,fldnms);
                }
            } else
                st=addrecord(fn,keystr,UPDFLDS,fldnms);
            break;
        case 'C' :
            st=mDB_createindex(fn,idxno,keystr);
            break;
        case 'd' :
            st=mDB_delete(fn,idxno,keystr);
            break;
        case 'e' :
            st=mDB_find(fn,idxno,  keystr,0);
            break;
        case 'f' :
            st=mDB_first(fn,idxno,NOROWS,0);
            break;
        case 'i' :
            st=mDB_header(fn,0,0);
            break;
        case 'l' :
            st=mDB_last (fn,idxno,NOROWS,0);
            break;
        case 'n' :
            st=mDB_next(fn,idxno,  keystr,NOROWS,0);
            break;
        case 'p' :
            st=mDB_prev(fn,idxno,  keystr,NOROWS,0);
            break;
        case 'r' :
            st=mDB_read(fn,(UL)atol(keystr),0);
            break;
        case 's' :
            st=mDB_search(fn,idxno,keystr,NOROWS,0);
            break;
        case 'u' :
            if (argc>3) {
                if (UPDFLDS[0]) {
                    char dta[513];
                    rebuilddata(argv[3],UPDFLDS,fldnms,EMPTYID,dta);
                    st=mDB_update(fn,idxno,keystr,EMPTYID,dta,0);
                } else
                    st=mDB_update(fn,idxno,keystr,EMPTYID,argv[3],0);
            }
            break;
        case 'B' :
            st=mDB_rebuildindex(fn,idxno);
            break;
        case 'D' :
            st=mDB_dropindex(fn,idxno);
            break;
        case 'L' :
            st=mDB_log(fn,keystr);
            break;
        default  :
            printf("unknown cmd [%c]\n",argv[1][0]);
            syntax();
            break;
        }
        if (st!=1) printf("Error %d\n",st);
        mDB_close(fn);
    }
    mDB_exit();
    exit(st);
}

